AsyncDisplayKit/Private/ASDisplayNode+FrameworkPrivate.h

```diff
@@ -0,0 +1,109 @@
+/* Copyright (c) 2014-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+//
+// The following methods are ONLY for use by _ASDisplayLayer, _ASDisplayView, and ASDisplayNode.
+// These methods must never be called or overridden by other classes.
+//
+
+#import "_ASDisplayLayer.h"
+#import "_AS-objc-internal.h"
+#import "ASDisplayNodeExtraIvars.h"
+#import "ASDisplayNode.h"
+#import "ASSentinel.h"
+#import "ASThread.h"
+#import "ASLayoutOptions.h"
+
+/**
+ Hierarchy state is propogated from nodes to all of their children when certain behaviors are required from the subtree.
+ Examples include rasterization and external driving of the .interfaceState property.
+ By passing this information explicitly, performance is optimized by avoiding iteration up the supernode chain.
+ Lastly, this avoidance of supernode traversal protects against the possibility of deadlocks when a supernode is
+ simultaneously attempting to materialize views / layers for its subtree (as many related methods require property locking)
+ 
+ Note: as the hierarchy deepens, more state properties may be enabled.  However, state properties may never be disabled /
+ cancelled below the point they are enabled.  They continue to the leaves of the hierarchy.
+ */
+
+typedef NS_OPTIONS(NSUInteger, ASHierarchyState)
+{
+  /** The node may or may not have a supernode, but no supernode has a special hierarchy-influencing option enabled. */
+  ASHierarchyStateNormal       = 0,
+  /** The node has a supernode with .shouldRasterizeDescendants = YES.
+      Note: the root node of the rasterized subtree (the one with the property set on it) will NOT have this state set. */
+  ASHierarchyStateRasterized   = 1 << 0,
+  /** The node or one of its supernodes is managed by a class like ASRangeController.  Most commonly, these nodes are
+      ASCellNode objects or a subnode of one, and are used in ASTableView or ASCollectionView.
+      These nodes also recieve regular updates to the .interfaceState property with more detailed status information. */
+  ASHierarchyStateRangeManaged = 1 << 1,
```

Comment:
Ah, very good question.  Yeah, I'm pretty unhappy with the name RangeManaged.  I think the best idea would be to rename it, making it more high level, but it's possible that there are reasons to do otherwise.  If we need the granularity, it may be best to add two new bits to specify those states but still have a high level one that indicates whether anything will be "driving" .interfaceState.